sdk/python/kfp/components/_structures.py

```diff
@@ -13,473 +13,512 @@
-            struct['optional'] = self.optional
-        if self.pattern:
-            struct['pattern'] = self.pattern
-        
-        return struct
-    
-    def __repr__(self):
-        return self.__class__.__name__ + '.from_struct(' + str(self.to_struct()) + ')'
+from .modelbase import ModelBase
 
 
-class InputSpec(InputOrOutputSpec):
-    pass
+PrimitiveTypes = Union[str, int, float, bool]
+PrimitiveTypesIncludingNone = Optional[PrimitiveTypes]
 
 
-class OutputSpec(InputOrOutputSpec):
-    pass
+class InputSpec(ModelBase):
+    def __init__(self,
+        name: str,
+        type: Optional[Union[str, Dict, List]] = None,
+        description: Optional[str] = None,
+        default: Optional[PrimitiveTypes] = None,
+        optional: Optional[bool] = False,
+    ):
+        super().__init__(locals())
 
 
-class ContainerSpec:
-    def __init__(self, image:str, command:List=None, args:List=None, file_outputs:Mapping[str,str]=None):
-        if not isinstance(image, str):
-            raise ValueError('image must be a string')
-        self.image = image
-        self.command = command
-        self.args = args
-        self.file_outputs = file_outputs
-    
-    @staticmethod
-    def from_struct(spec_dict:Mapping):
-        spec_dict = copy.deepcopy(spec_dict)
-        
-        image = spec_dict.pop('image')
-        
-        container_spec = ContainerSpec(image)
-        
-        if 'command' in spec_dict:
-            container_spec.command = list(spec_dict.pop('command'))
-        if 'args' in spec_dict:
-            container_spec.args = list(spec_dict.pop('args'))
-        if 'fileOutputs' in spec_dict:
-            container_spec.file_outputs = dict(spec_dict.pop('fileOutputs'))
-
-        if spec_dict:
-            raise ValueError('Found unrecognized properties: {}'.format(spec_dict))
-        
-        return container_spec
-
-    def to_struct(self):
-        struct = OrderedDict()
-        if self.image:
-            struct['image'] = self.image
-        if self.command:
-            struct['command'] = self.command
-        if self.args:
-            struct['args'] = self.args
-        if self.file_outputs:
-            struct['fileOutputs'] = self.file_outputs
-        
-        return struct
-    
-    def __repr__(self):
-        return self.__class__.__name__ + '.from_struct(' + str(self.to_struct()) + ')'
+class OutputSpec(ModelBase):
+    def __init__(self,
+        name: str,
+        type: Optional[Union[str, Dict, List]] = None,
+        description: Optional[str] = None,
+    ):
+        super().__init__(locals())
 
 
-def check_instance_type(obj, accepted_types:List[type]):
-    if not [accepted_type for accepted_type in accepted_types if isinstance(obj, accepted_type)]:
-        raise ValueError('Encountered object of wrong type. Accepted types: {}. Actual type: {}'.format(accepted_types, obj.__class__.__name__))
+class InputValuePlaceholder(ModelBase): #Non-standard attr names
+    _original_names = {
```

Comment:
"_original_names" are confusing to me. Rename it to something like "serialized names"?

Reply:
Done.